### DISCUSSION TITLE
Use `winget upgrade PACKAGE_NAME` within NodeJS `cp.spawn()`

### DISCUSSION THREAD

**Discussion by onyx-nxt** [microsoft/winget-cli#2791] [Q&A]
When running the command `winget upgrade PACKAGE_NAME` within NodeJS `cp.spawn()`, the progress bar is going to stdout as parts of the process, not as a whole. How would I make it so I can detect every time progress has been made in downloading the file rather than after it's done downloading?

Current:
```js
[
'starting...',
'\b \r\r  ██████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  17.0 MB / 79.4 MB\r  ████████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒  43.0 MB / 79.4 MB\r  ████████████████████████▒▒▒▒▒▒  64.0 MB / 79.4 MB\r  ██████████████████████████▒▒▒▒  70.0 MB / 79.4 MB\r  █████████████████████████████▒  77.0 MB / 79.4 MB\r  ██████████████████████████████  79.4 MB / 79.4 MB\r\n',
'done!'
]
```

What I want:
```js
[
'starting...',
'\b \r\r  ██████▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒  17.0 MB / 79.4 MB',
'\r  ████████████████▒▒▒▒▒▒▒▒▒▒▒▒▒▒  43.0 MB / 79.4 MB',
'\r  ████████████████████████▒▒▒▒▒▒  64.0 MB / 79.4 MB',
'\r  ██████████████████████████▒▒▒▒  70.0 MB / 79.4 MB'
'\r  █████████████████████████████▒  77.0 MB / 79.4 MB\r'
'\r  ██████████████████████████████  79.4 MB / 79.4 MB\r\n',
'done!'
]
```

**Comment by denelon** [comment]
The progress bar takes input to calculate and render progress. It's using the CR without the LF to give the visual appearance of animated progress on the same "line" of output. You might be able to swap a CR without a LF into a CRLF.

**Reply by onyx-nxt** [reply]
How would I do that?

**Reply by Trenly** [reply]
You would need to watch the output as a stream and check for the `\r` character instead of waiting for individual elements to be added to the array

**Reply by onyx-nxt** [reply]
> You would need to watch the output as a stream and check for the `\r` character instead of waiting for individual elements to be added to the array
@Trenly 

So I would use `stdout.on('data', ...)` right?

**Reply by Trenly** [reply]
> So I would use `stdout.on('data', ...)` right?

No clue. I'm not a Node expert, and using child processes I know even less. All I know is that winget writes the progress bar to the output stream for the console. In order to do what you’re looking to do, you need to capture that output stream and look for the Carraige Return sequence instead of the Line Feed sequence

**Reply by onyx-nxt** [reply]
> No clue. I'm not a Node expert, and using child processes I know even less. All I know is that winget writes the progress bar to the output stream for the console. In order to do what you’re looking to do, you need to capture that output stream and look for the Carraige Return sequence instead of the Line Feed sequence

I'm not sure if it is possible with nodejs, is there a download file which i can track and get the progress while its downloading?
